doc/design.md

```diff
@@ -0,0 +1,184 @@
+# Introduction
+
+## Acknowledgments
+
+This document's section structure is derived from the "[arc42](https://arc42.org/)" architectural template by Dr. Gernot Starke, Dr. Peter Hruschka.
+
+## Terms and Abbreviations
+
+<dl>
+    <dt>BFSJ</dt><dd>BucketFS Java</dd>
+</dl>
+
+# Constraints
+
+## Log-based Synchronization Check has a Minimum Resolution of one Second
+`const~log-based-synchronization-check-has-a-minimum-resolution-of-one-second~1`
+
+At least the monitoring solution based on cluster logs is limited to a resolution of one second.
+
+That means this monitor cannot distinguish between subsequent uploads to the same object in a bucket if they are not at least one second apart.
+
+For more details on what this means and how we deal with this constraint see the design decision in section ["How do we validate that objects on BucketFs are ready to use"](#how-do-we-validate-that-objects-on-bucketfs-are-ready-to-use).
+
+Needs: dsn
+
+# Solution Strategy
+
+BucketFS offers a web API that is not compatible with established standards like [Web DAV](http://www.webdav.org/). While in some parts similar, the differences are big enough that standard client libraries can't be used. This library abstracts the underlying HTTP requests and responses to a level that lets users instead deal with buckets and their contents directly.
+
+## Requirement Overview
+
+Please refer to the [System Requirement Specification](system_requirements.md) for user-level requirements.
+
+# Building Blocks
+
+This section introduces the building blocks of the software. Together those building blocks make up the big picture of the software structure.
+
+## `Bucket*Configuration`
+
+The `Bucket*Configuration` is a set of objects representing the setup of an Exasol cluster.
+
+Needs: impl
+
+## `Bucket`
+
+The `Bucket` building block controls interaction with a bucket in BucketFS.
+
+# Runtime
+
+This section describes the runtime behavior of the software.
+
+## BucketFS Access
+
+### List of `Bucket`Contents
+`dsn~bucket-lists-its-contents~1`
+
+The `Bucket` lists its contents as a set of object names.
+
+Covers:
+
+* `req~bucket-content-listing~1`
+
+Needs: impl, itest
+
+### Uploading to `Bucket`
+`dsn~uploading-to-bucket~1`
+
+The `Bucket` offers uploading a file from a locally accessible filesystem to a bucket in BucketFS.
+
+Covers:
+
+* `req~uploading-a-file-to-bucketfs~1`
+
+Needs: impl, itest
+
+### Uploading Strings to `Bucket`
+`dsn~uploading-strings-to-bucket~1`
+
+The `Bucket` offers uploading strings into a file in bucket in BucketFS.
+
+Covers:
+
+* `req~uploading-text-to-a-file-in-bucketfs~1`
+
+Needs: impl, itest
+
+### Uploading InputStream to `Bucket`
+`dsn~uploading-input-stream-to-bucket~1`
+
+The `Bucket` offers uploading of the contents of an `InputStream` into a file in that bucket on BucketFS.
+
+Covers:
+
+* `req~uploading-input-stream-to-a-file-in-bucketfs~1`
+
+Needs: impl, itest
+
```

Comment:
I think these design declarations do not add any additional information. I would suggest to leave away design here and directly add `Needs: impl, itest` to the requirements.
In my opinion, this is currently a duplication of the requirements.